File: src/executor/binary-sut.ts
/**
 * Binary SUT Wrapper
 *
 * Enables arbitrary binaries (Python scripts, compiled executables, etc.)
 * to be used as System Under Test in ppef experiments.
 *
 * The BinarySut class implements the SUT interface and bridges to external
 * processes via stdin/stdout IPC, with proper timeout handling and error isolation.
 *
 * Supported I/O formats:
 * - json: Structured data serialization (default)
 * - raw: Plain text passthrough
 * - lines: Line-separated values
 */

import { spawn } from "node:child_process";

import type { SUT } from "../types/sut.js";

/**
 * Configuration for binary SUT execution.
 */
export interface BinarySutConfig {
	/** Command to execute (e.g., "python3", "./my-script", "node") */
	command: string;

	/** Arguments to pass to the command */
	args?: string[];

	/** Working directory (defaults to current directory) */
	cwd?: string;

	/** Environment variables (merged with process.env) */
	env?: Record<string, string>;

	/** How to serialize inputs to stdin */
	inputFormat?: "json" | "raw" | "lines";

	/** How to deserialize stdout */
	outputFormat?: "json" | "raw" | "lines";

	/** Timeout per run in milliseconds (0 = no timeout, default: 30000) */
	timeout?: number;

	/** Exit code that indicates success (default: 0) */
	successExitCode?: number;
}

/**
 * Result from binary process execution.
 */
interface ProcessResult {
	stdout: string;
	stderr: string;
	exitCode: number | null;
}

/**
 * Binary SUT wrapper implementing the SUT interface.
 *
 * Spawns external processes, communicates via stdin/stdout,
 * and provides timeout handling and error isolation.
 */
export class BinarySut implements SUT<unknown, unknown> {
	readonly id: string;
	readonly config: Readonly<BinarySutConfig>;

	constructor(id: string, config: BinarySutConfig) {
		this.id = id;
		this.config = {
			...config,
			inputFormat: config.inputFormat ?? "json",
			timeout: config.timeout ?? 30000,
			successExitCode: config.successExitCode ?? 0,
		};
	}

	/**
	 * Execute the binary with the given inputs.
	 *
	 * @param inputs - Algorithm inputs (will be serialized based on inputFormat)
	 * @returns Promise resolving to deserialized output
	 * @throws Error on timeout, non-zero exit code, or spawn failure
	 */
	async run(inputs: unknown): Promise<unknown> {
		const proc = spawn(this.config.command, this.config.args ?? [], {
			cwd: this.config.cwd,
			env: { ...process.env, ...this.config.env },
			stdio: ["pipe", "pipe", "pipe"],
		});

		// Serialize inputs based on format
		const inputData = this._serializeInputs(inputs);
		proc.stdin.write(inputData);
		proc.stdin.end();

		// Collect output with timeout
		const result = await this._collectOutput(proc);
		return this._deserializeOutput(result);
	}

	/**
	 * Serialize inputs to stdin format.
	 */
	private _serializeInputs(inputs: unknown): string {
		switch (this.config.inputFormat) {
			case "json":
				return JSON.stringify(inputs);
			case "raw":
				return String(inputs);
			case "lines":
				return Array.isArray(inputs) ? inputs.join("\n") + "\n" : String(inputs) + "\n";
			default:
				return JSON.stringify(inputs);
		}
	}

	/**
	 * Collect stdout/stderr from process with timeout handling.
	 */
	private async _collectOutput(proc: ReturnType<typeof spawn>): Promise<ProcessResult> {
		return new Promise((resolve, reject) => {
			let stdout = "";
			let stderr = "";

			proc.stdout?.on("data", (data: Buffer) => {
				stdout += data.toString();
			});

			proc.stderr?.on("data", (data: Buffer) => {
				stderr += data.toString();
			});

			const timeoutMs = this.config.timeout ?? 0;
			let timer: ReturnType<typeof setTimeout> | undefined;

			if (timeoutMs > 0) {
				timer = setTimeout(() => {
					proc.kill("SIGKILL");
					reject(new Error(`BinarySut timeout after ${timeoutMs}ms`));
				}, timeoutMs);
			}

			proc.on("close", (exitCode) => {
				if (timer) clearTimeout(timer);

				const successCode = this.config.successExitCode ?? 0;
				if (exitCode === successCode) {
					resolve({ stdout, stderr, exitCode });
				} else {
					reject(new Error(`BinarySut exited with code ${exitCode}: ${stderr}`));
				}
			});

			proc.on("error", (error: Error) => {
				if (timer) clearTimeout(timer);
				reject(new Error(`BinarySut failed to spawn: ${error.message}`));
			});
		});
	}

	/**
	 * Deserialize stdout to output format.
	 */
	private _deserializeOutput(output: ProcessResult): unknown {
		const trimmed = output.stdout.trim();
		if (!trimmed) return null;

		switch (this.config.outputFormat) {
			case "json":
				return JSON.parse(trimmed);
			case "raw":
				return trimmed;
			case "lines":
				return trimmed.split("\n").filter(Boolean);
			default:
				// Auto-detect: try JSON first, fall back to raw
				try {
					return JSON.parse(trimmed);
				} catch {
					return trimmed;
				}
		}
	}
}

/**
 * Options for creating a binary SUT factory.
 */
export interface CreateBinarySutOptions extends BinarySutConfig {
	/** SUT ID (defaults to command name) */
	id?: string;
}

/**
 * SUT factory type for binary SUTs.
 *
 * Matches the expected SutFactory signature from the SUT interface.
 */
export type BinarySutFactory = (config?: Record<string, unknown>) => SUT<unknown, unknown>;

/**
 * Create a binary SUT factory function.
 *
 * The returned factory matches the SutFactory signature and can be used
 * interchangeably with other SUT factories in the ppef framework.
 *
 * @example
 * ```typescript
 * import { createBinarySut } from "ppef";
 *
 * const pythonSutFactory = createBinarySut({
 *   id: "python-classifier",
 *   command: "python3",
 *   args: ["classifier.py"],
 *   inputFormat: "json",
 *   outputFormat: "json",
 *   timeout: 30000,
 * });
 *
 * const sut = pythonSutFactory({ modelPath: "./model.pkl" });
 * const result = await sut.run({ features: [1, 2, 3] });
 * ```
 */
export function createBinarySut(options: CreateBinarySutOptions): BinarySutFactory {
	return (config?: Record<string, unknown>) => {
		const mergedConfig: BinarySutConfig = {
			command: options.command,
			args: options.args,
			cwd: options.cwd,
			env: options.env,
			inputFormat: options.inputFormat,
			outputFormat: options.outputFormat,
			timeout: options.timeout,
			successExitCode: options.successExitCode,
			...config,
		};

		const id = (config?.id as string | undefined) ?? options.id ?? `binary-${options.command}`;
		return new BinarySut(id, mergedConfig);
	};
}

File: python/src/ppef/executor/binary_sut.py
"""Binary SUT Wrapper.

Enables arbitrary binaries (Python scripts, compiled executables, etc.)
to be used as System Under Test in PPEF experiments.

The BinarySUT class wraps subprocess execution for cross-language SUT
integration, communicating via stdin/stdout IPC with configurable
serialization formats and timeout handling.

Supported I/O formats:
- json: Structured data serialization (default)
- raw: Plain text passthrough
- lines: Line-separated values
"""

from __future__ import annotations

import json
import os
import subprocess
from dataclasses import dataclass, field
from typing import Any, Literal

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

InputFormat = Literal["json", "raw", "lines"]
OutputFormat = Literal["json", "raw", "lines"]


@dataclass(frozen=True)
class BinarySUTConfig:
    """Configuration for binary SUT execution."""

    command: str
    args: list[str] = field(default_factory=list)
    cwd: str | None = None
    env: dict[str, str] = field(default_factory=dict)
    input_format: InputFormat = "json"
    output_format: OutputFormat = "json"
    timeout: float = 30.0
    success_exit_code: int = 0


# ---------------------------------------------------------------------------
# BinarySUT
# ---------------------------------------------------------------------------


class BinarySUT:
    """Binary SUT wrapper implementing the SUT protocol.

    Spawns external processes, communicates via stdin/stdout,
    and provides timeout handling and error isolation.
    """

    def __init__(self, id: str, config: BinarySUTConfig) -> None:
        self._id = id
        self._config = config

    @property
    def id(self) -> str:
        return self._id

    @property
    def config(self) -> BinarySUTConfig:
        return self._config

    def run(self, inputs: Any) -> Any:
        """Execute the binary with the given inputs.

        Serializes *inputs* to stdin according to ``input_format``,
        runs the subprocess with the configured timeout, and
        deserializes stdout according to ``output_format``.

        Raises on timeout (SIGKILL), non-zero exit code, or spawn failure.
        """
        input_data = self._serialize_inputs(inputs)

        env = {**os.environ, **self._config.env} if self._config.env else None

        timeout_seconds = self._config.timeout if self._config.timeout > 0 else None

        try:
            proc = subprocess.run(
                [self._config.command, *self._config.args],
                input=input_data,
                capture_output=True,
                text=True,
                cwd=self._config.cwd,
                env=env,
                timeout=timeout_seconds,
            )
        except subprocess.TimeoutExpired:
            raise TimeoutError(f"BinarySUT timeout after {self._config.timeout}s") from None
        except FileNotFoundError:
            raise RuntimeError(
                f"BinarySUT failed to spawn: command '{self._config.command}' not found"
            ) from None

        if proc.returncode != self._config.success_exit_code:
            raise RuntimeError(f"BinarySUT exited with code {proc.returncode}: {proc.stderr}")

        return self._deserialize_output(proc.stdout)

    # ------------------------------------------------------------------
    # Serialization
    # ------------------------------------------------------------------

    def _serialize_inputs(self, inputs: Any) -> str:
        match self._config.input_format:
            case "json":
                return json.dumps(inputs)
            case "raw":
                return str(inputs)
            case "lines":
                if isinstance(inputs, list):
                    return "\n".join(str(item) for item in inputs) + "\n"
                return str(inputs) + "\n"

    def _deserialize_output(self, stdout: str) -> Any:
        trimmed = stdout.strip()
        if not trimmed:
            return None

        match self._config.output_format:
            case "json":
                return json.loads(trimmed)
            case "raw":
                return trimmed
            case "lines":
                return [line for line in trimmed.split("\n") if line]


# ---------------------------------------------------------------------------
# Factory
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class CreateBinarySUTOptions:
    """Options for creating a binary SUT factory."""

    command: str
    id: str | None = None
    args: list[str] = field(default_factory=list)
    cwd: str | None = None
    env: dict[str, str] = field(default_factory=dict)
    input_format: InputFormat = "json"
    output_format: OutputFormat = "json"
    timeout: float = 30.0
    success_exit_code: int = 0


def create_binary_sut(
    options: CreateBinarySUTOptions,
) -> Any:
    """Create a binary SUT factory function.

    The returned callable matches the SUT factory signature and can be
    used interchangeably with other SUT factories in the PPEF framework.

    Example::

        factory = create_binary_sut(CreateBinarySUTOptions(
            id="python-classifier",
            command="python3",
            args=["classifier.py"],
            input_format="json",
            output_format="json",
            timeout=30.0,
        ))

        sut = factory({"model_path": "./model.pkl"})
        result = sut.run({"features": [1, 2, 3]})
    """

    def factory(config: dict[str, Any] | None = None) -> BinarySUT:
        merged = {
            "command": options.command,
            "args": list(options.args),
            "cwd": options.cwd,
            "env": dict(options.env),
            "input_format": options.input_format,
            "output_format": options.output_format,
            "timeout": options.timeout,
            "success_exit_code": options.success_exit_code,
        }

        if config is not None:
            merged.update(config)

        sut_id = (
            (
                config.get("id")
                if isinstance(config, dict) and isinstance(config.get("id"), str)
                else None
            )
            or options.id
            or f"binary-{options.command}"
        )

        return BinarySUT(
            id=sut_id,
            config=BinarySUTConfig(
                command=merged["command"],
                args=merged["args"],
                cwd=merged["cwd"],
                env=merged["env"],
                input_format=merged["input_format"],
                output_format=merged["output_format"],
                timeout=merged["timeout"],
                success_exit_code=merged["success_exit_code"],
            ),
        )

    return factory
